src/btree/bt_split.c

```diff
@@ -1509,30 +1517,47 @@ __split_multi_inmem(WT_SESSION_IMPL *session, WT_PAGE *orig, WT_MULTI *multi, WT
  *     Discard moved update lists from the original page.
  */
 static void
-__split_multi_inmem_final(WT_PAGE *orig, WT_MULTI *multi)
+__split_multi_inmem_final(WT_SESSION_IMPL *session, WT_PAGE *orig, WT_MULTI *multi)
 {
     WT_SAVE_UPD *supd;
+    WT_UPDATE *upd;
     uint32_t i, slot;
 
+    /* If we have saved updates, we must have decided to restore them to the new page. */
+    WT_ASSERT(session, multi->supd_entries == 0 || multi->supd_restore);
+
     /*
      * We successfully created new in-memory pages. For error-handling reasons, we've left the
      * update chains referenced by both the original and new pages. We're ready to discard the
      * original page, terminate the original page's reference to any update list we moved.
      */
-    for (i = 0, supd = multi->supd; i < multi->supd_entries; ++i, ++supd)
-        switch (orig->type) {
-        case WT_PAGE_COL_FIX:
-        case WT_PAGE_COL_VAR:
+    for (i = 0, supd = multi->supd; i < multi->supd_entries; ++i, ++supd) {
+        /* We have finished restoration. Discard the update chains that aren't restored. */
+        if (!supd->restore)
+            continue;
+
+        if (supd->ins == NULL) {
+            slot = WT_ROW_SLOT(orig, supd->ripcip);
+            upd = orig->modify->mod_row_update[slot];
+            orig->modify->mod_row_update[slot] = NULL;
+        } else {
+            upd = supd->ins->upd;
             supd->ins->upd = NULL;
-            break;
-        case WT_PAGE_ROW_LEAF:
-            if (supd->ins == NULL) {
-                slot = WT_ROW_SLOT(orig, supd->ripcip);
-                orig->modify->mod_row_update[slot] = NULL;
-            } else
-                supd->ins->upd = NULL;
-            break;
         }
+
+        /*
+         * Free the onpage value and the older versions moved to the history store. We can't free
+         * the updates for in memory database and fixed length column store as they don't support
+         */
+        if (supd->onpage_upd != NULL && !F_ISSET(S2C(session), WT_CONN_IN_MEMORY) &&
+          orig->type != WT_PAGE_COL_FIX) {
+            WT_ASSERT(session, upd != NULL && upd != supd->onpage_upd);
+            for (; upd->next != NULL && upd->next != supd->onpage_upd; upd = upd->next)
```

Comment:
I find this loop confusing (I got all the way through asking why we were truncating after the on page value, before I figured out that `upd->next` is the outcome). Can you switch it so the logic is based on `upd`, instead of `upd->next`?

Something like:
`for (upd = upd->next; upd != NULL && upd != supd->onpage_upd; upd = upd->next)`

Reply:
That would be incorrect. We need the pointer to point at the update before the onpage value so we can free everything from the onpage value. 

`for (upd = upd->next; upd != NULL && upd != supd->onpage_upd; upd = upd->next)` make upd points to the onpage value. In that case, we can't delink onpage value from the update chain.

Reply:
__wt_free_update_list(session, &upd->next); also set upd->next to NULL.

Reply:
If we track upd and set upd to NULL does not move the onpage value from the update chain.

Reply:
If that does help understanding, can you give code examples of the section instead of just that line.

Reply:
I think:

```
for (upd = upd->next; upd != NULL && upd != supd->onpage_upd; upd = upd->next)
    ;
WT_ASSERT(session, upd == supd->onpage_upd);
__wt_free_update_list(session, &upd);
```
Should be equivalent. The code doesn't use `upd` again, so having it set to `NULL` by `__wt_free_update_list` should be OK.

Reply:
The problem here is the address of onpage value is still linked to the update chain even it's freed.

V4 -> V3 (onpage) -> V2 -> V1

We need to set V4->next to NULL. So we need to move the pointer to V4 and set V4->next to NULL in __wt_free_update_list(session, &upd->next);

Reply:
Set upd to NULL only clears the temporary pointer. It doesn't truncate the onpage value from the update chain.

Reply:
I'm not following. `upd` is a pointer so has exactly the same value as `upd->next` would in your version of the code. It's not a different pointer to the same memory location, it's the same pointer.
If by temporary pointer you mean getting a reference because we are passing by address, that's reversed when `__wt_free_update_list` does `*updp = NULL`;

Did you try this change and it doesn't work? If so, let's get on a video call some time this afternoon to get on the same page.

Reply:
@bvpvamsikrishna would you mind taking a look at this? I feel like I'm missing something, but I can't see what it is.

Reply:
upd lives on the stack but prev_upd->next is in the heap. Am I correct? So setting upd to NULL will not affect the value of prev_upd->next on the heap.

Reply:
I'm trying your change.

Reply:
Tests are failing.

Reply:
May be that's more understandable?
```
for (; upd != NULL && upd->next != supd->onpage_upd; upd = upd->next)
    ;
WT_ASSERT(session, upd->next == supd->onpage_upd);
tmp = upd->next;
upd->next = NULL;
__wt_free_update_list(session, &tmp);
```

Reply:
Yes - I your suggested new version is better. If I saw the code as it was I'd try to restructure it once I understood - so I think something clearer is better.

Reply:
Changed.